### PR TITLE
new interface

### PR DESCRIPTION
**Pull request by charlierudolph**
@alexdavid @kevgo 

This is my proposed new interface for mycha.
chai, sinon, and sinon-chai would no longer be dependencies.
If we come to an agreement on this, I will move forward with implementing it.

### CODE REVIEW

**Comment by charlierudolph** [comment]
@alexdavid @kevgo thoughts?

File: README.md [hunk 1]
```diff
@@ -10,8 +14,57 @@
 
 
 ## Usage
-* Tests must end in `_(spec|test).(coffee|js)`
-* Run tests with ```mycha run```
+
+* `mycha install <testHelperPath>`
```
Comment (kevgo) [inline README.md]: Nitpicking here, but it should probably query the test helper path from the user. This command is run infrequent enough that people don't remember its syntax.

File: README.md [hunk 2]
```diff
@@ -10,8 +14,57 @@
...
+    sinon = require 'sinon'
+    chai.use require 'sinon-chai'
+
+    global.expect = chai.expect
+    global.sinon = sinon
+
+    process.env.NODE_ENV = 'test'
+    ```
+
+  * writes to `mycha.coffee`
```
Comment (kevgo) [inline README.md]: This could also be a JSON or CSON file

**Comment by kevgo** [comment]
I like it! I think is better to make the test setup that Mycha provides explicit in the code base. This way its easier for app developers to maintain and extend it.

**Comment by kevgo** [comment]
Seriously cool Readme-driven development btw!  :+1:

File: README.md (hunk 2 above)
Comment (charlierudolph) [inline README.md]: Yeah, I'd prefer to just use a coffee file for now since there is a regular expression in here though.
Can easily add support for other forms of configs later

**Comment by charlierudolph** [comment]
Alright this is ready to roll. For finding tests, ended up walking the tree by hand because `glob` is slower since it traverses the node_modules folder.

File: README.md [hunk 3]
```diff
@@ -10,8 +14,53 @@
...
+    module.exports =
+
+      # Default options to pass to mocha (can be overriden by command line options)
+      mochaOptions:
+        colors: yes
+        compilers: 'coffee:coffee-script/register'
+        reporter: 'dot'
+
+      # Regular expressions for files that are tests
+      testFiles: [
```
Comment (kevgo) [inline README.md]: Only make this an array if there are multiple elements in here. For now this can just be

```coffeescript
testFiles: /_(spec|test)\\.(coffee|js)$/
```

File: README.md [hunk 4]
```diff
@@ -10,8 +14,53 @@
...
+        /_(spec|test)\\.(coffee|js)$/
+      ]
+
+      # Files to include before all tests
+      testHelpers: [
```
Comment (kevgo) [inline README.md]: Same here, remove the array

File: features/install.feature [hunk 5]
```diff
@@ -0,0 +1,45 @@
+Feature: install
+
+  Scenario:
+    Given I am in a node project
+    When I run "mycha install"
```
Comment (kevgo) [inline features/install.feature]:
```cucumber
When I run "might install" entering
   | spec/spec_helper.coffee |
```

File: features/install.feature [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    And I enter "spec/spec_helper.coffee"
+    And I wait for the command to exit
+
+    Then I now have the following devDependencies
```
Comment (kevgo) [inline features/install.feature]:
```cucumber
Then Mycha has added the devDependencies "chai", "sinon", and "sinon-chai" to my package.json file
```

File: features/install.feature [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+      | MODULE     |
+      | chai       |
+      | sinon      |
+      | sinon-chai |
+    And I now have a the file "spec/spec_helper.coffee" with the contents
```
Comment (kevgo) [inline features/install.feature]: Remove the first "the"

File: features/run_tests.feature [hunk 8]
```diff
@@ -0,0 +1,52 @@
+Feature: run tests
+
+  Background:
+    Given I run "mycha install"
```
Comment (kevgo) [inline features/run_tests.feature]: The particularities of setting up Mycha aren't relevant for this test setup.

```cucumber
Given I have a standard Mycha installation
```

File: features/run_tests.feature [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    And I enter "spec/spec_helper.coffee"
+    And I wait for the command to exit
+
+
+  Scenario: no tests
+    Given I have no test files
+    When I run "mycha"
+    Then I see the output "0 passing"
```
Comment (kevgo) [inline features/run_tests.feature]:
```cucumber
Then I see "0 passing"

File: features/run_tests.feature [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+
+
+  Scenario: passing test
+    Given I have the file "one_spec.coffee" which passes
```
Comment (kevgo) [inline features/run_tests.feature]:
```cucumber
Given I have a spec file with 1 passing spec

File: features/run_tests.feature [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+    When I run "mycha"
+    Then I see the output "1 passing"
+
+
+  Scenario: failing test
+    Given I have the file "one_spec.coffee" which fails
```
Comment (kevgo) [inline features/run_tests.feature]: Given I have a spec file with 1 failing spec

File: features/run_tests.feature [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+  Scenario: running file
+    Given I have the file "one_spec.coffee" which passes
+    And I have the file "dir/two_spec.coffee" which passes
+    When I run "mycha one_spec.coffee"
+    Then I see the output "1 passing"
+
+
+  Scenario: running folder
+    Given I have the file "one_spec.coffee" which passes
```
Comment (kevgo) [inline features/run_tests.feature]: This file specifies 3 different things:
* the output for various passing and failing specs
* how spec files are found/globbed
* how individual spec files can be provided via the command line

I would separate those into separate files, and use scenario outlines there.

File: features/step_definitions/file_steps.coffee [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+  @Given /^I have no test files$/, (done) ->
+    done() # No op for readability
+
+
+  @Given /^I have the file "([^"]*)" which passes$/, (filePath, done) ->
+    fsExtra.outputFile(
+      path.join @tmpDir, filePath
+      '''
+      it 'passes', ->
+        expect(true).to.be.true
```
Comment (kevgo) [inline features/step_definitions/file_steps.coffee]: This line is not necessary

File: lib/argument_builder.coffee [hunk 14]
```diff
@@ -0,0 +1,48 @@
+_ = require 'lodash'
+fs = require 'fs'
+minimist = require 'minimist'
+path = require 'path'
+TestFinder = require './test_finder'
+
+
+class ArgumentBuilder
```
Comment (kevgo) [inline lib/argument_builder.coffee]: Please add a comment for this class

File: lib/cli.coffee [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+    if err then return done err
+    done null, testHelperPath
+
+
+installMycha = (testHelperPath, done) ->
+  installer = new MychaInstaller {cwd, testHelperPath}
+  installer.install done
+
+
+if minimist(argv)._[0] is 'install'
```
Comment (kevgo) [inline lib/cli.coffee]: Not a fan of minimist. Why can't you use nconf here?

File: lib/mocha_runner.coffee [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+class MochaRunner
+
+  constructor: ({argv, cwd}) ->
+    @argumentBuilder = new ArgumentBuilder {argv, cwd}
+
+
+  run: (done) ->
+    @argumentBuilder.build (err, args) ->
+      if err then return done err
+      childProcess = spawn mochaExecutablePath, args, stdio: 'inherit'
```
Comment (kevgo) [inline lib/mocha_runner.coffee]: I would try to use the Mocha JS API here, instead of forking a subprocess: https://github.com/mochajs/mocha/blob/master/test/jsapi/index.js

File: lib/mycha_installer.coffee [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+path = require 'path'
+
+
+class MychaInstaller
+
+  constructor: ({@cwd, @testHelperPath}) ->
+
+
+  install: (done) ->
+    async.parallel [
```
Comment (kevgo) [inline lib/mycha_installer.coffee]: I would run those in series. If the `npm install` fails, don't do the other steps. This also allows you to output what exactly you have done, so that the developer knows (see attached file). 
__ _fish

File: lib/mycha_installer.coffee [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+      @_writeMychaConfig
+      @_writeTestHelper
+    ], done
+
+
+  _installDependencies: (done) =>
+    exec "npm install --save-dev chai sinon sinon-chai", {@cwd}, done
+
+
+  _writeMychaConfig: (done) =>
```
Comment (kevgo) [inline lib/mycha_installer.coffee]: I would use Gulp or something like that to generate these files. The installation workflow should be:
* npm install -g mycha
* mycha install
  * adds `mycha` to the devDependencies
  * creates all the files
  * runs `npm install`

File: lib/mycha_installer.coffee [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+  install: (done) ->
+    async.parallel [
+      @_installDependencies
+      @_writeMychaConfig
+      @_writeTestHelper
+    ], done
+
+
+  _installDependencies: (done) =>
+    exec "npm install --save-dev chai sinon sinon-chai", {@cwd}, done
```
Comment (kevgo) [inline lib/mycha_installer.coffee]: Use the JS API of NPM here. An example is in the RMS code base.

File: lib/test_finder.coffee [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+    @_searchDirectory @cwd, [], done
+
+
+  _isTestFile: (filePath) ->
+    return yes for regex in @testFiles when filePath.match regex
+    no
+
+
+  _searchDirectory: (dir, testFiles, done) ->
+    fs.readdir dir, (err, files) =>
```
Comment (kevgo) [inline lib/test_finder.coffee]: Use a file walker like https://github.com/oleics/node-filewalker here.

File: lib/argument_builder.coffee [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+      if err then return done err
+      @_getTestFiles config, (err, files) =>
+        if err then return done err
+        done null, @_buildDefault(config).concat(@argv).concat(files)
+
+
+  _buildDefault: ({mochaOptions, testHelpers}) ->
+    args = []
+    for own key, value of mochaOptions
+      args.push "--#{key}"
```
Comment (kevgo) [inline lib/argument_builder.coffee]: In general, Node land has are a ton of libraries out there to do things like what you do here. Maybe you can use https://www.npmjs.com/package/dargs here?

File: lib/mocha_runner.coffee (hunk 16 above)
Comment (charlierudolph) [inline lib/mocha_runner.coffee]: We would have to do all the argument parsing ourselves then. 
Mocha doesn't have a API where you can pass in the argv, thats all done in the file located in bin.

File: lib/test_finder.coffee (hunk 20 above)
Comment (charlierudolph) [inline lib/test_finder.coffee]: k

File: lib/mycha_installer.coffee (hunk 19 above)
Comment (charlierudolph) [inline lib/mycha_installer.coffee]: k

File: lib/mocha_runner.coffee (hunk 16 above)
Comment (kevgo) [inline lib/mocha_runner.coffee]: Isn't minimist doing the arg parsing for you?

Comment (kevgo) [inline lib/mocha_runner.coffee]: ah, ok

File: README.md (hunk 1 above)
Comment (charlierudolph) [inline README.md]: done

File: features/install.feature (hunk 5 above)
Comment (charlierudolph) [inline features/install.feature]: done

File: features/install.feature (hunk 6 above)
Comment (charlierudolph) [inline features/install.feature]: done

File: features/install.feature (hunk 7 above)
Comment (charlierudolph) [inline features/install.feature]: removed the `a`

File: features/run_tests.feature (hunk 8 above)
Comment (charlierudolph) [inline features/run_tests.feature]: done

File: features/run_tests.feature (hunk 9 above)
Comment (charlierudolph) [inline features/run_tests.feature]: done

File: features/run_tests.feature (hunk 10 above)
Comment (charlierudolph) [inline features/run_tests.feature]: done

File: features/run_tests.feature (hunk 11 above)
Comment (charlierudolph) [inline features/run_tests.feature]: done

File: features/run_tests.feature (hunk 12 above)
Comment (charlierudolph) [inline features/run_tests.feature]: done

File: features/step_definitions/file_steps.coffee (hunk 13 above)
Comment (charlierudolph) [inline features/step_definitions/file_steps.coffee]: done

File: lib/cli.coffee (hunk 15 above)
Comment (charlierudolph) [inline lib/cli.coffee]: `nconf` uses `optimist` which the maintainer has deprecated because he uses `minimist`. 

Removed from here as what I'm doing is so simple I don't need minimist but still use it in ArgumentBuilder to determine if there are any files specified.

File: lib/mycha_installer.coffee (hunk 18 above)
Comment (charlierudolph) [inline lib/mycha_installer.coffee]: mycha currently isn't meant to be installed globally. It still has mocha as a dependency so you need it installed locally to run tests on ci.

Comment (charlierudolph) [inline lib/mycha_installer.coffee]: sorry I understand what you're doing now, I see no reason to have it installed globally

File: README.md (hunk 2 above)
Comment (kevgo) [inline README.md]: Isn't CSON normal executable CoffeeScript? That's at least what it is often criticized for.

File: lib/test_finder.coffee (hunk 20 above)
Comment (charlierudolph) [inline lib/test_finder.coffee]: ended up using a different walker but done

File: lib/mycha_installer.coffee (hunk 19 above)
Comment (charlierudolph) [inline lib/mycha_installer.coffee]: done

File: lib/mycha_installer.coffee (hunk 18 above)
Comment (kevgo) [inline lib/mycha_installer.coffee]: Okay, I forgot that Mycha can be installed locally and added to the dev dependencies with a single command as well. 

The advantage of installing Mycha globally is that it would make adding Mycha to a lot of repos easier, because all you have to do is `mycha install` instead of `npm install --save-dev mycha`

File: lib/argument_builder.coffee (hunk 21 above)
Comment (charlierudolph) [inline lib/argument_builder.coffee]: cool

File: README.md (hunk 2 above)
Comment (charlierudolph) [inline README.md]: I personally like coffee files. 
The only difference in the file is I need to add `modules.exports`
The difference is use is that I can require the file instead of needing another module to parse it.

File: README.md (hunk 3 above)
Comment (charlierudolph) [inline README.md]: I don't think the added flexibility is worth the extra complexity

File: README.md (hunk 2 above)
Comment (kevgo) [inline README.md]: Agreed that this can be added later

File: lib/mocha_runner.coffee [hunk 22]
```diff
@@ -0,0 +1,19 @@
+MochaArgumentBuilder = require './mocha_argument_builder'
+mochaExecutablePath = require.resolve 'mocha/bin/mocha'
+{spawn} = require 'child_process'
```
Comment (kevgo) [inline lib/mocha_runner.coffee]: Consider using https://www.npmjs.com/package/cross-spawn here so that this runs on Windows as well. This is open source.

File: lib/mycha_installer.coffee (hunk 18 above)
Comment (charlierudolph) [inline lib/mycha_installer.coffee]: In terms of gulp generating files, I've never seen it used for scaffolding. Do you have an example or know of what module to use? Its also for only two files so unless its simple to set up, I'd prefer just doing it by hand

File: lib/mycha_installer.coffee (hunk 17 above)
Comment (charlierudolph) [inline lib/mycha_installer.coffee]: updated to series, are you suggesting the output also includes what files are written.
That may look weird after the `npm install`. The current output is:
```
sinon-chai@2.7.0 ../../../../var/folders/yn/rz7n_fdn2vq5_rdmvn06yn500000gn/T/tmp-4827gEMsTw2ADKOx/node_modules/sinon-chai

chai@2.1.2 ../../../../var/folders/yn/rz7n_fdn2vq5_rdmvn06yn500000gn/T/tmp-4827gEMsTw2ADKOx/node_modules/chai
├── assertion-error@1.0.0
└── deep-eql@0.1.3 (type-detect@0.1.1)

sinon@1.14.1 ../../../../var/folders/yn/rz7n_fdn2vq5_rdmvn06yn500000gn/T/tmp-4827gEMsTw2ADKOx/node_modules/sinon
├── lolex@1.1.0
├── util@0.10.3 (inherits@2.0.1)
└── formatio@1.1.1 (samsam@1.1.2)
```

**Comment by charlierudolph** [comment]
@kevgo all comments addressed.
@alexdavid can I also please get your take?

File: lib/mycha_installer.coffee (hunk 18 above)
Comment (kevgo) [inline lib/mycha_installer.coffee]: What I want to see here is a more industrial scale solution for installing things into an existing repository. There are a number of edge cases that need to be properly handled by Mycha if we want it to be usable in serious projects. Like what happens if the test helper already exists. The user should have the option to
* keep his existing test helper
* replace the existing one with the new one
* have the installer merge the required changes into the existing test helper
* see a diff between the existing and the new helper to do the changes himself
* abort the install safely

I think the installer should also display which exact files and directories it has created, similar to how Rails generators work, so that the user knows what just happened.

There are (of course) libraries that provide such functionality. In Ruby I would use [Thor](https://github.com/erikhuda/thor). I haven't found a direct counterpart for that in Node. But there is [Yeoman](http://yeoman.io) (Grunt based) and [Slush](http://slushjs.github.io/#/) (Gulp based). 

Please check out Yeoman, since they have really nailed the UX part of this.
* npm install --global yeoman yo generator-node
* mkdir test ; cd test
* yo node
  * check out how comfortably it displays the options, default values, and allows to modify them
* run it again, and play with the different options (press "d" to show a diff if there are conflicts etc)

Thor also allows to [add content into existing files](https://github.com/kevgo/sections_rails/blob/master/lib/generators/sections_generator.rb#L9-L11). We might need that in case the spec helper already exists.

All of this can happen in a separate PR.

Comment (charlierudolph) [inline lib/mycha_installer.coffee]: Okay, yeah will look at generators in a follow up PR. 
For now are we okay with it just overwriting the file if it exists? 
Should I add some output so the user knows that the files are written?
```shell
$ mycha install
prompt: Default test helper path: spec/spec_helper.coffee

npm install chai sinon sinon-chai # in bold
# npm install output

create spec/spec_helper.coffee
create mycha.coffee
```

Comment (kevgo) [inline lib/mycha_installer.coffee]: You could add a simple "created <filename>" message to the console for now. That should be pretty easy. Bonus points for making it green using https://github.com/sindresorhus/chalk

**Comment by charlierudolph** [comment]
@kevgo updated, @alexdavid still waiting on a review from you
Here is a sample output

---------------------------------------
screen shot 2015-03-26 at 7 56 58 pm
----------------------------------------

File: README.md (hunk 3 above)
Comment (alexdavid) [inline README.md]: I think it's confusing to have regexes in an array because it's not obvious if it finds the intersection or the union here. I think an array of strings, or just a single regex makes more sense

File: lib/test_finder.coffee [hunk 23]
```diff
@@ -0,0 +1,24 @@
...
+
+
+class TestFinder
+
+  constructor: ({@cwd, @testFiles}) ->
+
+
+  find: (done) ->
+    glob(
+      "#{@cwd}/**/*"
```
Comment (alexdavid) [inline lib/test_finder.coffee]: I think glob if faster at finding matches if this is just `"#{@cwd}/**"`

**Comment by alexdavid** [comment]
This looks good to me.  A rewrite of mycha has been long over due. Thanks for doing this Charlie!

File: README.md (hunk 3 above)
Comment (charlierudolph) [inline README.md]: Okay. I think a single regex should suffice.

File: lib/test_finder.coffee (hunk 23 above)
Comment (charlierudolph) [inline lib/test_finder.coffee]: done

**Comment by charlierudolph** [comment]
@alexdavid all comments addressed

**Comment by charlierudolph** [comment]
@alexdavid @kevgo any last comments or is this good to merge? Just want to double check since this is such a major change.

**Comment by kevgo** [comment]
Good from my end

**Comment by alexdavid** [comment]
Looks good to me. Let's ship it
